### PR TITLE
Core typing: sendrecv.py

### PR DESCRIPTION
**Pull request by gpotter2**
This PR:
- adds core typing to `sendrecv.py`

~~**Needs: https://github.com/secdev/scapy/pull/3055**~~

This is one of the last extremly big & annoying PRs of this kind. There are several parts left but none this central.

### CODE REVIEW

**Review by gpotter2 — COMMENTED** [review]
This is ready for review. Among other things, this PR discovered:
- fix "Bad number of arguments" type of issues on `PcapReader` & cie.
- type overloading of some utils: remove several cast
- fix *args being misplaced (x2)
- add some doc for sendrecv functions
- fix `select_func` being called with inconsistent types (dict or lists)
- fix an edge case in `AsyncSniffer` which could result in `self.thread.start()` being called with `self.thread = None`
- remove unused variables in some `sendrecv` functions
- *remove compat.py from sphinx generation: see the commit for an explanation. Basically there's a builtin docstring in the `typing` module that is wrong, and that we can't fix.*

~~**This however still requires #3055 to be merged first**~~

**Review by polybassa — COMMENTED** [review]
I'll carefully read through this PR. I couldn't find any issue.

**Comment by gpotter2** [comment]
@guedou @p-l- Hi, this is ready for review. (btw It appears we're currently having an issue in our tests related to OpenSSL versions.)

**Review by p-l- — APPROVED** [review]
LGTM, great work again!

**Comment by gpotter2** [comment]
@guedou Your call

File: scapy/utils.py [hunk 1]
```diff
@@ -159,7 +163,19 @@ def __eq__(self, other):
         return super(EDecimal, self).__eq__(other) or float(self) == other
 
 
-def get_temp_file(keep=False, autoext="", fd=False):
+@overload
+def get_temp_file(keep, autoext, fd):
+    # type: (bool, str, Literal[True]) -> IO[bytes]
+    pass
```
[the remark this answers is not in the document]
Comment (guedou) [reply scapy/utils.py:169]: Agree!

**Review by guedou — APPROVED** [review]

**Pull request merged**